### ISSUE TITLE
IR generation crashes the compiler when a public function calls a function that instantiates a contract using new

### ISSUE THREAD

**Issue by cameel** [ethereum/solidity#9992]
## Steps to Reproduce
Using the `--ir` option when the contract has a public function that calls another function (not necessarily public) that just uses `new` to create a contract instance results in a crash due to the compiler trying to access a non-existent key in a map:

```
Unknown exception during compilation: map::at
```

And it's not even an `InternalCompilerError` but just a crash on an unknown exception.

This does not happen if `new` is removed or replaced with something else. Putting `new` directly in the public function without the intermediate call doesn't cause a crash either.

### Example 1
```solidity
contract A {}

contract B {
    function f() public {
        new A();
    }
}

contract C is B {
    function g() public {
        f();
    }
}
```

### Example 2
```solidity
contract A {}

function f() {
    new A();
}

contract C {
    function g() public {
        f();
    }
}
```

### Compiling
```bash
solc test.sol --ir
```

## Environment
- Compiler version: 0.7.3+commit.9bfce1f6.Linux.g++

**Comment by ekpyron** [comment]
It's ``IRGenerator.cpp:L88``. Apparently we don't track which contracts are created from a contract indirectly through the call chain.
[For the record: I just checked what happens and won't look into fixing it right away, so feel free to pick it up]

**Comment by cameel** [comment]
Also, it goes back to the time contract creation calls were implemented in IR. It crashes on 0.6.7 too. 0.6.6 reports `UnimplementedFeatureError` instead:
```
Unimplemented feature:
/solidity/libsolidity/codegen/ir/IRGeneratorForStatements.cpp(758): Throw in function virtual void solidity::frontend::IRGeneratorForStatements::endVisit(const solidity::frontend::FunctionCall&)
Dynamic exception type: boost::wrapexcept<solidity::langutil::UnimplementedFeatureError>
std::exception::what: FunctionKind 7 not yet implemented
[solidity::util::tag_comment*] = FunctionKind 7 not yet implemented
```

**Comment by chriseth** [comment]
Does it mean that `_contract.annotation().contractDependencies` does not contain the correct information?

**Comment by Marenz** [comment]
Example 1 is fixed by #10030

**Comment by Marenz** [comment]
With the current changes Example 2 is also fixed by #10030

**Comment by Marenz** [comment]
fixed by #10030
